RUBY-1563 Implement polling SRV records for mongos discovery

saghm opened the pull request.

p-mongo commented on lib/mongo/srv/monitor.rb:
```diff
@@ -0,0 +1,92 @@
+        begin
+          @records = @resolver.get_records(@records.hostname)
+        rescue Resolv::ResolvTimeout => e
+          log_warn("Timed out trying to resolve hostname #{@records.hostname}")
```
Please also include the exception class and message in these warnings.

p-mongo commented on lib/mongo/srv/monitor.rb:
```diff
@@ -0,0 +1,92 @@
+# limitations under the License.
+
+module Mongo
+  module SRV
```
Since we decided to camel case module names like this, `Srv`?

saghm replied:
This would be inconsistent with existing user-facing APIs, which we can't change without breaking backwards compatibility: https://github.com/mongodb/mongo-ruby-driver/blob/e3d399d758cc8fa780c02c49c66b8cf4f9700bd6/lib/mongo/error/no_srv_records.rb#L24

p-mongo commented on lib/mongo/cluster/topology/sharded.rb:
```diff
@@ -27,6 +27,15 @@ class Sharded < Base
+        def initialize(options, monitoring, cluster)
+          super(options, monitoring, cluster)
+
+          unless options[:srv_records].nil?
```
This option should be documented.

p-mongo commented on lib/mongo/srv/monitor.rb:
```diff
@@ -0,0 +1,92 @@
+module Mongo
+  module SRV
+
+    class Monitor
```
This and all other classes need to either be marked api private or have docstrings written for them.

p-mongo commented:
There need to be tests written that adding and removing hosts via srv polling publishes correct sdam events. In particular, there must be topology change events published which I don't believe will currently happen when removing hosts.

Per the spec language, the hosts should be added first and removed second.

https://github.com/mongodb/specifications/blob/master/source/polling-srv-records-for-mongos-discovery/polling-srv-records-for-mongos-discovery.rst

p-mongo commented on lib/mongo/srv/records.rb:
```diff
@@ -0,0 +1,66 @@
+        self
+      end
+
+      def validate_record!(record_host, hostname)
```
While I see spec tests for various host names, I specifically would like to also see unit tests for this method when it is given `record_host` and `host_name` with and without trailing dots. For example, something like `foo.bar.` and `bar`.

Can you please also add a docstring explaining what this method is supposed to do as it is impossible to tell from the signature.

saghm replied:
This method is copied verbatim from Mongo::URI::SRVProtocol; it implements the requirement from the initial seedlist discovery that the driver should error when the SRV record returns a record with a different domain name (in the fifth paragraph of [this section](https://github.com/mongodb/specifications/blob/master/source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.rst#seedlist-discovery):

> Drivers MUST raise an error and MUST NOT initiate a connection to any returned host name which does not share the same {domainname}.

p-mongo commented:
@saghm Can you please create a separate PR with this change only:

> I abstracted some of the SRV polling functionality previously in Mongo::URI::SRVProtocol into the Mongo::SRV module to facilitate reuse in the monitoring

p-mongo commented:
Per https://github.com/mongodb/mongo-ruby-driver/pull/1280, `monitoring_io: false` should result in the srv monitoring being omitted.

I am not seeing where in the diff the monitors are closed. Killing threads in finalizers is insufficient.

Also, monitors are attached to topologies but we consider topologies lightweight (as well as immutable) and replace them any time a server description changes, for example. This means 1) adding a server to a sharded topology would recreate the monitor, possibly creating some difficult to diagnose interactions, and 2) there are going to be a ton of monitors alive for any given client that does experience multiple sdam events.

p-mongo commented:
The spec also requires that srv rescans are done for unknown topologies.

p-mongo commented:
Server addition and removal in the SDAM flow is done under the SDAM flow lock, therefore when adding/removing servers as part of srv monitoring these operations either need to be done under the same lock or there needs to be an explanation added why this locking is not required.

saghm commented:
> Per the spec language, the hosts should be added first and removed second.

You linked to the entire spec, so I'm not sure which part you're talking about. I don't see anything specifying the order of hosts being added and removed, and I don't see why it would matter.

> I am not seeing where in the diff the monitors are closed. Killing threads in finalizers is insufficient.

What do you suggest instead? Killing a thread to stop it from running seems pretty standard to me.

p-mongo commented:
The spec says:

<pre>
For all verified host names, as returned through the DNS SRV query, the driver:

    MUST add each valid new host to the topology as Unknown
    MUST remove all hosts that are part of the topology, but are no longer in the returned set of valid hosts
</pre>

Furthermore, we currently consider a topology with no servers (addresses) in it a terminal state. Therefore:

1. The logic that reports that the empty topology is a terminal state should be audited to see if it needs to be updated for SRV monitoring. Since SRV monitoring must happen for unknown topologies, it seems to me that an empty topology is now a state that the driver can transition out of.
2. There need to be tests added that make the topology empty and then non-empty via SRV monitoring.

saghm commented:
PR for abstracting SRV behavior: https://github.com/mongodb/mongo-ruby-driver/pull/1285
